src/windows/View.tsx

```diff
@@ -0,0 +1,367 @@
+        }
+
+        return childContext;
+    }
+
+    setFocusRestricted(restricted: boolean) {
+        if (!this._focusManager || !this.props.restrictFocusWithin) {
+            console.error('View: setFocusRestricted method requires restrictFocusWithin property to be set to true');
+            return;
+        }
+
+        if (restricted) {
+            this._focusManager.restrictFocusWithin();
+        } else {
+            this._focusManager.removeFocusRestriction();
+        }
+    }
+
+    setFocusLimited(limited: boolean) {
+        if (!this._focusManager || !this.props.limitFocusWithin) {
+            console.error('View: setFocusLimited method requires limitFocusWithin property to be set to true');
+            return;
+        }
+
+        if (limited && !this._isFocusLimited) {
+            this._isFocusLimited = true;
+            this._focusManager.limitFocusWithin();
+        } else if (!limited && this._isFocusLimited) {
+            this._isFocusLimited = false;
+            this._focusManager.removeFocusLimitation();
+        }
+    }
+
+    private _onFocusableKeyDown = (e: React.SyntheticEvent<any>): void => {
+
+        let keyEvent = EventHelpers.toKeyboardEvent(e);
+        if (this.props.onKeyPress) {
+            this.props.onKeyPress(keyEvent);
+        }
+
+        if (this.props.onPress) {
+            let key = keyEvent.keyCode;
+            // ENTER triggers press on key down
+            if (key === KEY_CODE_ENTER) {
+                this.props.onPress(keyEvent);
+            }
+        }
+    }
+
+    private _onFocusableKeyUp = (e: React.SyntheticEvent<any>): void => {
+
+        let keyEvent = EventHelpers.toKeyboardEvent(e);
+        if (keyEvent.keyCode === KEY_CODE_SPACE) {
+            if (this.props.onPress) {
+                this.props.onPress(keyEvent);
+            }
+        }
+    }
+
+    private _onFocus = (e: React.SyntheticEvent<any>): void => {
+        this.onFocus();
+        if (this.props.onFocus) {
+            this.props.onFocus(EventHelpers.toFocusEvent(e));
+        }
+    }
+
+    private _onBlur = (e: React.SyntheticEvent<any>): void => {
+        if (this.props.onBlur) {
+            this.props.onBlur(EventHelpers.toFocusEvent(e));
+        }
+    }
+
+    // From FocusManagerFocusableComponent interface
+    //
+    onFocus() {
+        // Focus Manager hook
+    }
+
+    getTabIndex(): number | undefined {
+        // Focus Manager may override this
+        return this.props.tabIndex;
+    }
+
+    updateNativeTabIndex(): void {
+        if (this._focusableElement) {
+            let tabIndex: number = this.getTabIndex() || 0;
+            let windowsTabFocusable: boolean = tabIndex >= 0;
+
+            this._focusableElement.setNativeProps({
+                tabIndex: tabIndex,
+                isTabStop: windowsTabFocusable
+            });
+        }
+    }
+}
+
+// A value for tabIndex (even <0) marks a View as being potentially keyboard focusable
```

Comment:
Why is it focusable with a `tabIndex` of `-1`? That seems different than the web implementation.

What does "keyboard focusable" mean? That you can hit the tab key to focus the View?

Reply:
"keyboard focusable" is meant to be used when differentiating from "mouse focusable".
Otherwise, we use "tab focusable" for being able to keyboard focus by tabbing.

Now, I decided to keep the few views with tabIndex==-1 under the FocusManager supervision in order to be able to implement the "focus the tabIndex==-1 component" trick. I'll take a new look, with all the changes you guys are asking for.

---
In reply to: [159586940](https://github.com/Microsoft/reactxp/pull/404#discussion_r159586940) [](ancestors = 159586940)